File: src/formosa/geomorphology/drainage/network/validation.py
"""
Validates flow graphs and report invalid topology.

Last modified: 2026-08-24, En-Chi Lee (williameclee@gmail.com)
"""

import numpy as np
from numpy.typing import NDArray

import formosa.geomorphology.drainage.network._backends.validation_py as val_py
from formosa.geomorphology._native import network_validation as val_f
from formosa.geomorphology.drainage.directions import DirectionEncoding
from formosa.geomorphology.drainage.neighbours import (
    compute_downstream_indices,
)
from formosa.utils import Backend, NpCanonIndex, raise_fortran_error
from formosa.utils.typing import NpCoords, NpFlowDir, NpIndex


class GraphTopologyError(RuntimeError):
    """
    Base exception for a graph that fails topology validation.
    """


class InvalidOriginalGraphTopology(GraphTopologyError):
    """
    Raised when an invalid result originated from invalid input topology.
    """


class UnresolvedSimplificationTopology(GraphTopologyError):
    """
    Raised when simplification leaves invalid topology from valid input.
    """


class DirectedFlowCycleError(GraphTopologyError):
    """
    Raised when the selected flow field contains one or more directed cycles.
    """

    def __init__(self, cycle_ijs: NDArray[np.integer]) -> None:
        self.cycle_ijs = np.asarray(cycle_ijs, dtype=np.int32).copy()
        super().__init__(
            f"Selected flow graph contains directed cycles at {self.cycle_ijs.tolist()}."
        )


class IncompleteFlowGraphError(GraphTopologyError):
    """
    Raised when construction omits one or more selected directed edges.
    """

    def __init__(
        self,
        missing_vtxs: NDArray[np.integer],
        missing_edges: NDArray[np.integer] | None = None,
    ) -> None:
        self.missing_ijs = np.asarray(missing_vtxs, dtype=np.int32).copy()
        if missing_edges is None:
            self.missing_edges = np.empty((0, 4), dtype=np.int32)
        else:
            self.missing_edges = np.asarray(missing_edges, dtype=np.int32).copy()
        super().__init__(
            f"Flow-graph construction omitted selected directed edges {self.missing_edges.tolist()}; "
            + f"participating cells are {self.missing_ijs.tolist()}."
        )


def _valid_flow_edges(
    dirs: NDArray[NpFlowDir], valids: NDArray[np.bool_], dir_enc: DirectionEncoding
) -> tuple[NDArray[NpCanonIndex], NDArray[NpCanonIndex], NDArray[np.bool_]]:
    """
    Returns downstream indices and a mask indicating whether the
    cell flows into a valid neighbouring (non-self) edge.
    """
    dsi, dsj, _, ds_inbounds = compute_downstream_indices(
        dirs, dir_enc, check=False, return_flat_index=False, oob_is_okay=True
    )

    # Whether the downstream cell is also valid (not just inbound)
    ds_valids = np.zeros(dirs.shape, dtype=bool)
    ds_valids[ds_inbounds] = valids[dsi[ds_inbounds], dsj[ds_inbounds]]

    # Exclude self-loops (where offsets di, dj == 0)
    not_self = dirs != dir_enc.no_flow_code

    has_valid_ds = valids & ds_valids & not_self
    return dsi, dsj, has_valid_ds


def _validate_flowgraph_coverage(
    vtxs: NDArray[NpIndex],
    arc_endpts: NDArray[NpIndex],
    dsi: NDArray[NpIndex],
    dsj: NDArray[NpIndex],
    has_valid_ds: NDArray[np.bool_],
) -> None:
    """
    Checks that every selected directed edge occurs in a returned
    graph arc.
    """
    represented = np.zeros(has_valid_ds.shape, dtype=bool)

    # Identify consecutive vertex pairs that belong to an arc.
    seg_cnts = np.zeros(vtxs.shape[0], dtype=np.int32)
    np.add.at(seg_cnts, arc_endpts[:, 0], 1)
    np.add.at(seg_cnts, arc_endpts[:, 1], -1)
    seg_valids = np.cumsum(seg_cnts)[:-1] > 0

    srcs = vtxs[:-1][seg_valids]
    targets = vtxs[1:][seg_valids]

    # Confirm each represented edge matches the source cell's expected downstream.
    matches = (targets[:, 0] == dsi[srcs[:, 0], srcs[:, 1]]) & (
        targets[:, 1] == dsj[srcs[:, 0], srcs[:, 1]]
    )
    matched_srcs = srcs[matches]
    represented[matched_srcs[:, 0], matched_srcs[:, 1]] = True

    missing_sources = np.argwhere(has_valid_ds & ~represented)
    if missing_sources.size:
        missing_targets = np.column_stack(
            (
                dsi[missing_sources[:, 0], missing_sources[:, 1]],
                dsj[missing_sources[:, 0], missing_sources[:, 1]],
            )
        )
        missing_edges = np.column_stack((missing_sources, missing_targets))
        missing_ijs = np.unique(missing_edges.reshape(-1, 2), axis=0)
        raise IncompleteFlowGraphError(missing_ijs, missing_edges)


def _locate_invalid_graph_topology_fortran(
    vtxs: NDArray[NpCoords],
    endpts: NDArray[NpIndex],
) -> NDArray[np.int32] | None:
    """
    Returns every topology violation using the capacity-aware
    Fortran scanner.

    The first scan uses a small provisional output buffer. If the
    exact count reported by that scan exceeds the buffer, the scan
    is repeated with a buffer of exactly the required size.
    Incomplete provisional results are never returned.

    Parameters
    ----------
    vtxs : NDArray[number]
        Vertex coordinates with shape `(V,2)`.
    endpts : NDArray[integer]
        Inclusive, 0-based arc endpoint indices with shape `(A,2)`.

    Returns
    -------
    NDArray[int32] or None
        Complete `(nintxs, 5)` intersection records using 0-based
        indices, or `None` when no violations are found.
    """
    vtxs_f = np.asfortranarray(vtxs.T, dtype=np.float32)
    endpts_f = np.asfortranarray(endpts.T, dtype=np.int32) + 1
    cpty = max(vtxs_f.shape[1] // 100, 3)  # Arbitrary capacity that seems to work

    intxs, nintxs, err_code = val_f.scan_invalid_graph_topology(vtxs_f, endpts_f, cpty)
    raise_fortran_error("scan_invalid_graph_topology", err_code)

    if nintxs == 0:
        return None

    if nintxs > cpty:
        exp_nintxs = nintxs
        intxs, nintxs, err_code = val_f.scan_invalid_graph_topology(
            vtxs_f, endpts_f, exp_nintxs
        )
        raise_fortran_error("scan_invalid_graph_topology", err_code)
        if nintxs != exp_nintxs:
            raise RuntimeError(
                "Topology-intersection count changed during exact-size retry."
            )

    intxs = intxs[:, :nintxs]
    intxs[:-1, :] -= 1  # Convert to 0-based indexing, except the intersection flag
    return intxs.T.astype(np.int32, order="C")


def locate_invalid_graph_topology(
    vtxs: NDArray[NpCoords], endpts: NDArray[NpIndex], backend: Backend = "fortran"
) -> NDArray[np.int32] | None:
    """
    Locates invalid topologies (segment intersections) within and
    between arcs in a graph.

    This function checks for self-intersections within individual
    arcs, as well as intersections between segments of different
    arcs. The intersection checks are performed using a 2D line
    segment intersection algorithm.

    Parameters
    ----------
    vtxs : NDArray[number]
        Grid coordinates (i, j) of each vertex.
        - Expected shape: `(nvtxs, 2)`.
    endpts : NDArray[int]
        Start and end vertex indices for each arc in `vtxs`.
        - Expected shape: `(narcs, 2)`.
    backend : {'fortran', 'python'}, optional
        Backend to use for computation.
        `'fortran'` uses the Fortran extension for performance,
        while `'python'` uses a pure Python implementation.
        - Default backend is `'fortran'`.

    Returns
    -------
    intxs : NDArray[int32] | None
        Detected intersections, or `None` if no intersections are
        found.
        - Shape: `(nintxs, 5)`.
        The rows are sorted lexicographically and each row contains:
        - `iarc`: Index of the first arc (0-based).
        - `jarc`: Index of the second arc (0-based).
        - `iseg`: Start vertex index of the first intersecting
            segment (0-based).
        - `jseg`: Start vertex index of the second intersecting
            segment (0-based).
        - `intx_flag`: Flag indicating the type of intersection:
            - 1 : Interior-interior crossing (X).
            - 2 : Collinear overlap, not identical.
            - 3 : Identical segment.
            - 4 : Endpoint-on-interior (T-junction).
            - 5 : Degenerate segment (some line is actually a point).
    """
    if vtxs.ndim != 2 or vtxs.shape[1] != 2:
        raise ValueError("Invalid array shapes passed.")
    elif endpts.ndim != 2 or endpts.shape[1] != 2:
        raise ValueError("Invalid array shapes passed.")

    match backend:
        case "python":
            intxs = val_py.locate_invalid_graph_topology(
                endpts.astype(np.int32, order="C"),
                vtxs.astype(np.float64, order="C"),
            )
            if not intxs:
                return None
            intxs = np.array(intxs, dtype=np.int32, order="C")
        case "fortran":
            intxs = _locate_invalid_graph_topology_fortran(vtxs, endpts)
            if intxs is None:
                return None
    if intxs.shape[0] > 1:
        # Sort lexicographically
        sort_idx = np.lexsort((intxs[:, 3], intxs[:, 2], intxs[:, 1], intxs[:, 0]))
        intxs = intxs[sort_idx]
    return intxs


def _ignore_identical_intergraph_arcs(
    intxs: NDArray[np.int32] | None,
    vtxs: NDArray[NpCoords],
    endpts: NDArray[NpIndex],
    graph_ids: NDArray[np.integer],
) -> NDArray[np.int32] | None:
    """
    Removes topology violations between identical arcs in different
    graphs.
    """
    if intxs is None:
        return None

    keeps = np.ones(intxs.shape[0], dtype=bool)
    identical_pairs: dict[tuple[int, int], bool] = {}
    for i, (iarc, jarc, _, _, _) in enumerate(intxs):
        if graph_ids[iarc] == graph_ids[jarc]:
            continue

        pair = (int(iarc), int(jarc))
        if pair not in identical_pairs:
            istart, iend = endpts[:, iarc]
            jstart, jend = endpts[:, jarc]
            iarc_vtxs = vtxs[:, istart : iend + 1]
            jarc_vtxs = vtxs[:, jstart : jend + 1]
            identical_pairs[pair] = np.array_equal(
                iarc_vtxs, jarc_vtxs
            ) or np.array_equal(iarc_vtxs, jarc_vtxs[:, ::-1])
        if identical_pairs[pair]:
            keeps[i] = False

    if not np.any(keeps):
        return None
    return intxs[keeps]


def _locate_disallowed_graph_topology(
    vtxs: NDArray[NpCoords],
    endpts: NDArray[NpIndex],
    graph_ids: NDArray[np.integer] | None = None,
) -> NDArray[np.int32] | None:
    """
    Locates violations in arrays stored in internal (2,N) layout.
    """
    intxs = locate_invalid_graph_topology(vtxs.T, endpts.T)
    if graph_ids is not None:
        intxs = _ignore_identical_intergraph_arcs(intxs, vtxs, endpts, graph_ids)
    return intxs

File: src/formosa/geomorphology/drainage/network/_backends/validation.f90
!> Validates flow-graph topology using the Fortran backend.
!!
!! This internal module is called by the Python network API and
!! other Fortran routines and is not intended to be used directly.
!!
!! Last modified: 2026-08-17, En-Chi Lee (williameclee@gmail.com)
module network_validation
    use utils, only: ERR_NO_ERROR, ERR_INVALID_INPUT, &
                     ERR_ALLOCATION_FAILURE
    use intersections, only: lines_intersect
    implicit none(type, external)
    private :: argsort_arcs, record_topology_intersection
contains
    pure function argsort_arcs(bboxes) result(indices)
        ! Helper function for 'scan_invalid_graph_topology' to sort the arcs by the left edge of their bounding box.
        implicit none(type, external)
        ! Arguments
        real, intent(in), contiguous :: bboxes(:, :)
        ! Outputs
        integer :: indices(size(bboxes, 2))
        ! Local variables
        integer :: i, j, h, index

        do i = lbound(bboxes, 2), ubound(bboxes, 2)
            indices(i) = i
        end do

        ! Shell sort idx by the left edge
        h = ubound(bboxes, 2)/2
        do while (h > 0)
            do i = h + 1, ubound(bboxes, 2)
                index = indices(i)
                j = i
                do while (j > h)
                    if (bboxes(1, indices(j - h)) > bboxes(1, index)) then
                        indices(j) = indices(j - h)
                        j = j - h
                    else
                        exit
                    end if
                end do
                indices(j) = index
            end do
            h = h/2
        end do
    end function argsort_arcs

    pure subroutine record_topology_intersection(record, intxs, nintxs)
        !! Counts one detected topology violation and stores it if capacity remains.
        !!
        !! The total count is incremented even after 'intxs' is full. This lets
        !! the caller distinguish the number stored from the exact number found
        !! and retry with an exactly sized buffer when necessary.
        implicit none(type, external)
        integer, intent(in) :: record(5)
            !! Intersection record: arc IDs, segment IDs, and intersection flag
        integer, intent(inout), contiguous :: intxs(:, :)
            !! Output buffer containing up to 'size(intxs, 2)' records
        integer, intent(inout) :: nintxs
            !! Total number of violations encountered, including unstored ones

        nintxs = nintxs + 1
        if (nintxs <= size(intxs, 2)) intxs(:, nintxs) = record
    end subroutine record_topology_intersection

    !> Scans all candidate segment pairs and returns the total
    !! violation count.
    !!
    !! Only the first 'capacity' violations are stored in 'intxs'.
    subroutine scan_invalid_graph_topology( &
        vtxs, arc_endpts, capacity, intxs, nintxs, err_code)
        implicit none(type, external)
        ! Arguments
        real, intent(in), contiguous :: vtxs(:, :)
            !! Vertex coordinates arranged as (2, V).
        integer, intent(in), contiguous :: arc_endpts(:, :)
            !! Inclusive, one-based arc endpoint indices arranged as '(2, narcs)'
        integer, intent(in) :: capacity
            !! Maximum number of intersection records that can be stored
        ! Outputs
        integer, intent(out) :: intxs(5, capacity)
            !! Stored intersection records; only the first
            !! 'min(nintxs, capacity)' columns are defined
        integer, intent(out) :: nintxs
            !! Exact number of violations found, which may exceed 'capacity'
        integer, intent(out) :: err_code
            !! Code indicating the status of the result
            !!   - 0: Programme executed properly
            !!   - 1: Input dimensions are incorrect, or input capacity is invalid
            !!   - 2: Memory allocation failed
        ! Local variables
        integer :: narcs
        integer :: i, j, iarc, jarc, iseg, jseg
        integer :: intx_flag, alloc_stat
        real, allocatable :: arc_bboxes(:, :)
        integer, allocatable :: idx(:)

        nintxs = 0
        err_code = ERR_NO_ERROR

        if (size(arc_endpts, 1) /= 2) then
            err_code = ERR_INVALID_INPUT
            return
        else if (size(vtxs, 1) /= 2) then
            err_code = ERR_INVALID_INPUT
            return
        else if (capacity < 1) then
            err_code = ERR_INVALID_INPUT
            return
        end if
        narcs = size(arc_endpts, 2)

        if (narcs == 0) return

        ! Construct the bounding boxes for each arc
        allocate (arc_bboxes(4, narcs), stat=alloc_stat)
        if (alloc_stat /= 0) then
            err_code = ERR_ALLOCATION_FAILURE
            return
        end if
        do iarc = 1, narcs
            arc_bboxes(1, iarc) = minval(vtxs(1, arc_endpts(1, iarc):arc_endpts(2, iarc)))
            arc_bboxes(2, iarc) = minval(vtxs(2, arc_endpts(1, iarc):arc_endpts(2, iarc)))
            arc_bboxes(3, iarc) = maxval(vtxs(1, arc_endpts(1, iarc):arc_endpts(2, iarc)))
            arc_bboxes(4, iarc) = maxval(vtxs(2, arc_endpts(1, iarc):arc_endpts(2, iarc)))
        end do

        ! Check arcs against themselves first
        do iarc = 1, narcs
            if (arc_endpts(2, iarc) - arc_endpts(1, iarc) == 1) cycle ! Skip if arc is just a single segment
            do iseg = arc_endpts(1, iarc), arc_endpts(2, iarc) - 1
            do jseg = iseg + 1, arc_endpts(2, iarc) - 1
                intx_flag = lines_intersect( &
                            vtxs(:, iseg), vtxs(:, iseg + 1), &
                            vtxs(:, jseg), vtxs(:, jseg + 1))
                if (intx_flag > 0) then
                    call record_topology_intersection( &
                        [iarc, iarc, iseg, jseg, intx_flag], intxs, nintxs)
                end if
            end do
            end do
        end do

        allocate (idx(narcs), stat=alloc_stat)
        if (alloc_stat /= 0) then
            err_code = ERR_ALLOCATION_FAILURE
            return
        end if
        idx = argsort_arcs(arc_bboxes)

        ! Check every arc against each other
        do i = 1, narcs
            iarc = idx(i)
            jloop: do j = i + 1, narcs
                jarc = idx(j)

                ! Skip if min x of right arc is greater than max x of left arc
                if (arc_bboxes(1, jarc) > arc_bboxes(3, iarc)) exit jloop

                ! Inline fast overlap check (no min/max calls)
                if (arc_bboxes(1, iarc) > arc_bboxes(3, jarc) .or. &
                    arc_bboxes(3, iarc) < arc_bboxes(1, jarc) .or. &
                    arc_bboxes(2, iarc) > arc_bboxes(4, jarc) .or. &
                    arc_bboxes(4, iarc) < arc_bboxes(2, jarc)) cycle

                do iseg = arc_endpts(1, iarc), arc_endpts(2, iarc) - 1
                do jseg = arc_endpts(1, jarc), arc_endpts(2, jarc) - 1
                    intx_flag = lines_intersect( &
                                vtxs(:, iseg), vtxs(:, iseg + 1), &
                                vtxs(:, jseg), vtxs(:, jseg + 1))
                    if (intx_flag > 0) then
                        ! Sort by arc ID
                        if (iarc < jarc) then
                            call record_topology_intersection( &
                                [iarc, jarc, iseg, jseg, intx_flag], intxs, nintxs)
                        else
                            call record_topology_intersection( &
                                [jarc, iarc, jseg, iseg, intx_flag], intxs, nintxs)
                        end if
                    end if
                end do
                end do
            end do jloop
        end do
    end subroutine scan_invalid_graph_topology
end module network_validation
